Hide public usages of bigdecimal::BigDecimal behind a crate feature

muzarski opened the pull request:
Refs: https://github.com/scylladb/scylla-rust-driver/issues/771

## Changes
- introduced `value::CqlDecimal` type
- hidden public usages of `bigdecimal::Bigdecimal` behind crate feature.
- wrapped `CqlDecimal` with `CqlValue::Decimal`
- updated docs
- updated tests
- added docstrings

## Points to discuss
- how to implement `PartialEq` for `CqlDecimal`? Note that `CqlDecimal` makes use of `CqlVarint`, so the bytes normalization part can be delegated to `CqlVarint`. However, there is an additional issue, because there are multiple pairs (integer, scale) which can represent the same number. Take for example: `1e3` and `100e1`. I don't think it's trivial to implement the comparison without converting bytes (hexadecimal representation) to decimal number.
- Should I create some macros for code repetitions? There are some parts of code that are repeated for each version of the `bigdecimal` crate that unfortunately cannot be expressed using some generic function (e.g. usages of `BigDecimal::new()` constructor).

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I have provided docstrings for the public items that I want to introduce.
- [x] I have adjusted the documentation in `./docs/source/`.
- [x] I added appropriate `Fixes:` annotations to PR description.

Lorak-mmk commented on scylla-cql/src/frame/response/result.rs:
This comment is regarding commit `tests: extract decimal test cases .
The message is not clear imho. Please explain in more detail what you did and why

Lorak-mmk commented on docs/source/data-types/decimal.md:
```diff
@@ -1,5 +1,14 @@
 # Decimal
-`Decimal` is represented as [`bigdecimal::BigDecimal`](https://docs.rs/bigdecimal/0.2.0/bigdecimal/struct.BigDecimal.html)
+`Decimal` is represented as `value::CqlDecimal` or [`bigdecimal::BigDecimal`](https://docs.rs/bigdecimal/0.2.0/bigdecimal/struct.BigDecimal.html)
```
Can we link to newest version instead of 0.2?

Lorak-mmk commented:
>     * how to implement `PartialEq` for `CqlDecimal`? Note that `CqlDecimal` makes use of `CqlVarint`, so the bytes normalization part can be delegated to `CqlVarint`. However, there is an additional issue, because there are multiple pairs (integer, scale) which can represent the same number. Take for example: `1e3` and `100e1`. I don't think it's trivial to implement the comparison without converting bytes (hexadecimal representation) to decimal number.

I guess it would be useful for users who don't need `BigDecimal`. This change should be backwards compatible so if implementing it is too hard, I think we can open an issue about it and skip to for now.
How does `BigDecimal` implement it?

>     * Should I create some macros for code repetitions? There are some parts of code that are repeated for each version of the `bigdecimal` crate that unfortunately cannot be expressed using some generic function (e.g. usages of `BigDecimal::new()` constructor).

I don't think it's important right now. We can do this in the future when repetitions become problematic.

piodul commented on scylla-cql/src/frame/value.rs:
```diff
@@ -931,6 +931,20 @@ impl Value for i64 {
+    fn serialize(&self, buf: &mut Vec<u8>) -> Result<(), ValueTooBig> {
+        let (bytes, scale) = self.as_signed_be_bytes_slice_and_exponent();
+
+        let serialized_len: i32 = bytes.len().try_into().map_err(|_| ValueTooBig)?;
+
+        buf.put_i32(serialized_len + 4);
```
Looks like it was copied from `impl Value for BigDecimal`. I think we have a bug in there as well.

I think we should explicitly check that `serialized_len < i32::MAX - 4` once and return `ValueTooBig` in that case.

piodul commented on scylla-cql/src/frame/value.rs:
```diff
@@ -364,6 +364,22 @@ impl std::hash::Hash for CqlVarint {
+///
+/// Represented as a pair:
+/// - a [`CqlVarint`] value
+/// - 64-bit integer which determines the position of the decimal point
```
The native representation (i.e. how Scylla/Cassandra transfers data over the socket) uses `i32` here.

piodul commented on scylla-cql/src/frame/value.rs:
```diff
@@ -364,6 +364,22 @@ impl std::hash::Hash for CqlVarint {
+/// The library support (e.g. conversion from [`CqlValue`]) for these types is
+/// enabled via `bigdecimal-02`, `bigdecimal-03` and `bigdecimal-04` crate features.
+///
+/// # DB data format
+/// Notice that [constructors](CqlDecimal#impl-CqlDecimal)
+/// don't perform any normalization on the provided data.
+/// For more details, see [`CqlVarint`] documentation.
```
Not sure if I'm a fan of adding docstrings in a later commit. I think they should be introduced in the same commit as the item they document so that they provide more context for the reviewer when they go through the PR commit by commit.

muzarski commented:
v2: addressed review comments, apart from the [comment](https://github.com/scylladb/scylla-rust-driver/pull/922#discussion_r1472579904) regarding removing support for `bigdecimal` v0.2 and v0.3. Waiting for @Lorak-mmk opinion on this topic.

muzarski commented on scylla-cql/src/frame/value.rs:
```diff
@@ -365,6 +365,56 @@ impl std::hash::Hash for CqlVarint {
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct CqlDecimal {
+    int_val: CqlVarint,
+    scale: i64,
```
I used `type Error = <i64 as TryInto<i32>>::Error>` as an error type for `TryFrom`. Not sure if this is correct here. Should we maybe introduce a struct representing the error?

piodul replied:
Not sure either. If the error message is informative enough then maybe we can leave it as it is.

I suppose that the representation of `CqlDecimal` won't change in the future, so we don't need extensibility of the error. If newer versions change the representation of `BigDecimal` and there will be more ways to fail, then we will introduce an error, but right now maybe it's not necessary.

piodul commented:
By the way, you need to update `Cargo.lock.msrv`. There is an instruction by @Lorak-mmk in `CONTRIBUTING.md` on how to do that.

Lorak-mmk commented on examples/Cargo.toml:
```diff
@@ -10,7 +10,7 @@ futures = "0.3.6"
 rustyline = "9"
 rustyline-derive = "0.6"
-scylla = {path = "../scylla", features = ["ssl", "cloud", "chrono", "time", "num-bigint-03", "num-bigint-04"]}
+scylla = {path = "../scylla", features = ["ssl", "cloud", "chrono", "time", "num-bigint-03", "num-bigint-04", "bigdecimal-02"]}
```
Makes sense to drop older versions. In that case, the first commit of this PR should bump the version from 0.2 to 0.4

muzarski commented:
v3: removed support for bigdecimal v0.2 and v0.3

muzarski commented:
v4: updated Cargo.lock.msrv file

piodul commented on scylla-cql/src/frame/value.rs:
```diff
@@ -886,9 +886,13 @@ impl Value for BigDecimal {
 
         let serialized = value.to_signed_bytes_be();
-        let serialized_len: i32 = serialized.len().try_into().map_err(|_| ValueTooBig)?;
+        let serialized_len: i32 = serialized
+            .len()
+            .checked_add(4)
+            .ok_or(ValueTooBig)
+            .and_then(|x| x.try_into().map_err(|_| ValueTooBig))?;
```
Now that I think about it, I'm not sure whether it is really possible to trigger the issue that supposedly happens here. It would require `serialize` to take up nearly all or more than available memory in the system (i.e. larger than `usize::MAX - 4).

I'll leave up the decision up to you what to do with it - another option would be to document why it's fine to add 4 and not check for overflow.

muzarski replied:
I think it's better to leave the check anyway.

piodul commented on scylla/src/utils/pretty.rs:
```diff
@@ -42,7 +42,10 @@ where
-            CqlValue::Decimal(d) => write!(f, "{}", d)?,
+            CqlValue::Decimal(d) => {
+                let (bytes, scale) = d.as_signed_be_bytes_slice_and_exponent();
+                write!(f, "blobAsDecimal(0x{:04x}{:x})", scale, HexBytes(bytes))?
```
`scale` should be printed as four bytes, not four digits - there should be 8 digits in total.

Besides, the protocol spec requires the value to be serialized in big-endian. The way you print the value conforms to that, but it's a little non-obvious. Perhaps you could use [to_be_bytes](https://doc.rust-lang.org/std/primitive.i32.html#method.to_be_bytes) here?

piodul reviewed (commented):
I think that in general bumping MSRV should be done in the same commit where `Cargo.lock` is updated - technically speaking, the change is not bisectable under` min_rust` job, but I really doubt that anybody will do something like that so that is a non-issue.

muzarski commented:
v5: addressed review comments

Lorak-mmk commented:
> I think that in general bumping MSRV should be done in the same commit where `Cargo.lock` is updated - technically speaking, the change is not bisectable under` min_rust` job, but I really doubt that anybody will do something like that so that is a non-issue.

Maybe it's not important for bisectability, but I'd still like those commits to be squashed. Updating Cargo.lock isn't an unrelated change in this case, it is a required change because of .toml modification.
Also in the commits that bump our packages versions we update .toml and .lock in 1 commit so it would be good to be consistent here.

muzarski commented:
v6: squashed `Cargo.lock.msrv` commit with `bigdecimal` version bump commit

muzarski commented on docs/source/data-types/data-types.md:
```diff
@@ -26,7 +26,7 @@ Database types and their Rust equivalents:
 * `Timestamp` <----> `value::CqlTimestamp`, `chrono::DateTime<Utc>`, `time::OffsetDateTime`
 * `Duration` <----> `value::CqlDuration`
-* `Decimal` <----> `bigdecimal::Decimal`
+* `Decimal` <----> `value::CqlDecimal`, `bigdecimal::Decimal` (v0.2.0, v0.3 and v0.4)
```
true..

muzarski commented:
v7: removed mention about bigdecimal v0.2 and v0.3 from the docs

piodul commented on docs/source/data-types/decimal.md:
```diff
@@ -1,5 +1,14 @@
+
+To make use of `bigdecimal::Bigdecimal` type, user should enable `bigdecimal-04` crate feature.
+
+## value::CqlDecimal
```
Perhaps we should have separate examples for `BigDecimal` and `CqlDecimal`? We already have separate examples for each type for `time.md`, `date.md` and `timestamp.md` (but it looks like we didn't do it for `varint.md`).

piodul reviewed (commented):
One last thing - I don't think it fixes #771 entirely.

First and foremost, the `chrono`, `time` and `secrecy` features enable dependencies that do not have stable releases, but their names do not have the version of those dependencies in them.

Second, the issue is not just about types used for serialization. We have other pre-1.0 dependencies which may prevent us from stabilizing the API:

- `histogram` - our `Metrics` type contains a `Histogram`. We should not expose it publicly but rather reexport its functionality through `Metrics`' methods.
- `num_enum` - it defines a bunch of traits and we implement them for some of our public types (and `num_enum::TryFromPrimitiveError` appears in our `FrameError`). We could easily implement the code that the crate generates automatically ourselves.
- `openssl` - it's optional right now but the crates themselves are not stable. While those are only bindings to OpenSSL and are compatible with a wide range of versions (in case of OpenSSL they seem to be forwards-compatible), the openssl crate itself might turn out to have some security issues and maybe we will have to bump its version.
- `strum`, `strum_macros` - I believe they have a similar problem to `num_enum` and the solution is the same for them.
- `async-trait` - we use it for some of our public traits (`AddressTranslator`, `AuthenticatorProvider`, `AuthenticatorSession`). We may consider switching to recently stabilized async traits or change the methods to return boxed futures.

There were also some [concerns](https://github.com/scylladb/scylla-rust-driver/issues/771#issuecomment-1916832035) about other, >=1.0 dependencies.
